### ISSUE TITLE
Date Format issue with Idp's providing fractional time

### ISSUE THREAD

**Issue by Glocktober** [HENNGE/minisaml#4]
Azure AD and possibly others provide fractional seconds in the time stamp for SAML validation timestamps. This causes format exceptions in `saml_to_datetime` (internal/saml.py) as `DATE_TIME_FORMAT` (internal/constants.py) is defined as `"%Y-%m-%dT%H:%M:%SZ"`.  Fractional time would require `".%f"` - i.e.  `"%Y-%m-%dT%H:%M:%S.%fZ"`

I've validated this format works with Azure AD, 

To work with both I'd suggest internal/saml.py saml_to_datatime() be updated  - maybe:

```python3
def saml_to_datetime(s: str) -> datetime.datetime:
     format = DATE_TIME_FORMAT
     if '.' in s:
            format = DATE_TIME_FORMAT_FRACTIONAL
     return datetime.datetime.strptime(s, format)
```

with `DATE_TIME_FORMAT_FRACTIONAL` added to constants.py

Regards

**Comment by ojii** [comment]
@Glocktober thank you for reporting this issue. I've just released version 20.8b3 which should resolve this problem.

**Comment by Glocktober** [comment]
Jonas,
Thank you for the update.
I had other issues - I was hitting the idp initiated endpoint by mistake. Hence setting forceAuthen wasn't working and I wasn't getting my request ID back. minisaml was processing the responses just fine.
I switched to the actual SAML API endpoint on Azure AD, but everything broke.  Azure was failing, giving errors that it could not decode the request.  
Sure enough I could not get any online SAML decoders to turn the SAMLRequest into XML.
I did some digging and found these did not like the urlsafe_base64 output. Apparently this just changes ('+' and '-' are mapped to other characters.)  This is part of the request deflate/b64 ecoding in requests.py line 28.  
I changed this from base64.urlsafe_b64decode to base64.b64decode and Azure worked fine.  Also I could decode the URL request with online decoders.
I'm not sure if that will break anything with any other IDP's you may be working with. If I have an opportunity this week I'll try this against simplesaml.
I like the simple interface of minisaml - and especially like that it keeps me out of XML. I hope this project gets a lot of use.
RegardsBob

    On Thursday, August 27, 2020, 02:33:49 AM EDT, Jonas Obrist <notifications@github.com> wrote:  
 
 


@Glocktober thank you for reporting this issue. I've just released version 20.8b3 which should resolve this problem.
